### PR TITLE
Implement method to apply adversarial patch onto a real-world surface

### LINKED ISSUE #853: Extend mask feature in AdversarialPatch*.apply_patch to include perspective transformation
We would like to extend the method `apply_patch` of the Adversarial Patch attacks to place the patch on a predefined area in the image, similar to `mask`, but also to recognise and apply the perspective transformations of the patch required to maximise overlap. It should also be possible to apply similar but randomly selected transformations during training.

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request adds a method to apply the patch to coordinates after perspective transformation. These coordinates can be the coordinates of a real-world, quadratic surface in the targeted image.

Thank you to @yusong-tan for sharing the code!

Fixes #853 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/adversarial_patch/adversarial_patch.py [hunk 1]
```diff
@@ -169,6 +168,19 @@ def reset_patch(self, initial_patch_value: Optional[Union[float, np.ndarray]]) -
         """
         self._attack.reset_patch(initial_patch_value=initial_patch_value)
 
+    def insert_transformed_patch(self, x: np.ndarray, patch: np.ndarray, image_coords: np.ndarray):
+        """
+        Insert patch to image based on given or selected coordinates.
+
+        :param x: The images to insert the patch.
```
Comment (minhitbk) [inline art/attacks/evasion/adversarial_patch/adversarial_patch.py:175]: I think this function is only applied for a single image, so this description confuses me that the patch can be applied for a batch of images.

File: art/attacks/evasion/adversarial_patch/adversarial_patch_numpy.py [hunk 2]
```diff
@@ -578,3 +579,17 @@ def reset_patch(self, initial_patch_value: Optional[Union[float, np.ndarray]]) -
             self.patch = initial_patch_value
         else:
             raise ValueError("Unexpected value for initial_patch_value.")
+
+    @staticmethod
+    def insert_transformed_patch(x: np.ndarray, patch: np.ndarray, image_coords: np.ndarray):
+        """
+        Insert patch to image based on given or selected coordinates.
+
+        :param x: The images to insert the patch.
```
Comment (minhitbk) [inline art/attacks/evasion/adversarial_patch/adversarial_patch_numpy.py:588]: I think this function is only applied for a single image, so this description confuses me that the patch can be applied for a batch of images.

File: art/attacks/evasion/adversarial_patch/adversarial_patch_tensorflow.py [hunk 3]
```diff
@@ -495,3 +496,17 @@ def reset_patch(self, initial_patch_value: Optional[Union[float, np.ndarray]] =
             self._patch.assign(initial_patch_value)
         else:
             raise ValueError("Unexpected value for initial_patch_value.")
+
+    @staticmethod
+    def insert_transformed_patch(x: np.ndarray, patch: np.ndarray, image_coords: np.ndarray):
+        """
+        Insert patch to image based on given or selected coordinates.
+
+        :param x: The images to insert the patch.
```
Comment (minhitbk) [inline art/attacks/evasion/adversarial_patch/adversarial_patch_tensorflow.py:505]: I think this function is only applied for a single image, so this description confuses me that the patch can be applied for a batch of images.

**Review by minhitbk — APPROVED** [review]
Hi @beat-buesser, your code looks fine. I only have one small comment and one suggestion: that is the function `insert_transform_patched` in the end is refered to just one utility function, so it is not necessary to create an intermediate call in the common class and then the 2 other classes call the same function. We can do directly in the common class and therefore simplify the code. However, I respect your decision with your design if you think it is needed to keep it this way.

**Comment by beat-buesser** [comment]
Hi @minhitbk Thank you very much for your review! I have updated the docstrings. The current approach with the static methods aims to enable users to directly use the specific implementations.

**Pull request merged**